flink-clients/src/main/java/org/apache/flink/client/RemoteExecutor.java

```diff
@@ -207,14 +207,23 @@ public JobExecutionResult executePlanWithJars(JobWithJars program) throws Except
 				shutDownAtEnd = false;
 			}
 
-			try {
-				return client.run(program, defaultParallelism).getJobExecutionResult();
-			}
-			finally {
-				if (shutDownAtEnd) {
-					stop();
-				}
-			}
+			final JobClient jobClient = client.run(program, defaultParallelism);
+
+			jobClient.addFinalizer(
+				new Runnable() {
+					@Override
+					public void run() {
+						if (shutDownAtEnd) {
```

Comment:
The closure is fine, but by moving this out of the runnable, we can save to register `Runnables` if we don't need them.